src/Runner.Common/RunServer.cs

```diff
@@ -67,10 +67,39 @@ private void CheckConnection()
             }
         }
 
-        public Task<AgentJobRequestMessage> GetJobMessageAsync(string id)
+        public Task<AgentJobRequestMessage> GetJobMessageAsync(string id, CancellationToken cancellationToken)
         {
             CheckConnection();
-            return _taskAgentClient.GetJobMessageAsync(id);
+            var jobMessage = RetryRequest<AgentJobRequestMessage>(async () =>
+                                                    {
+                                                        return await _taskAgentClient.GetJobMessageAsync(id, cancellationToken);
+                                                    }, cancellationToken);
+            return jobMessage;
+        }
+
+        private async Task<T> RetryRequest<T>(Func<Task<T>> func,
+                                              CancellationToken cancellationToken,
+                                              int maxRetryAttemptsCount = 5
+                                             )
+        {
+            var retryCount = 0;
+            while (true)
```

Comment:
```suggestion
            while (!cancellationToken.IsCancellationRequested)
```
?

Reply:
I thought about it, if we put condition to while, C# compiler thinks that code after loop can be reached and force to write some `return` or `throw` there. 
So in case when cancellation is requested before start of the loop, probably, we will have to either:
1) throw `OperationCanceledException`or any other exception manually
or
2) return some "empty" value and create additional condition for empty value in Runner function.

I'd prefer not to write the condition and create extra logic after the loop.

@TingluoHuang What do you think?

Reply:
Let's leave as it is ❤️